tests/test_util.py

```diff
@@ -26,3 +28,44 @@ def test_batched_batch_size_greater_than_iterable():
 def test_batched_invalid_batch_size():
     with pytest.raises(ValueError, match="n must be at least one"):
         list(_batched([1, 2, 3], 0))
+
+
+class TestLakeFSUriPartRegexes:
+    @pytest.mark.parametrize(
+        "repo_name, valid",
+        [
+            ("my-repo", True),
+            ("@@repo", False),
+            ("", False),
+            ("a", False),
+            ("a" * 63, True),
+            ("a" * 64, False),
+        ],
+    )
+    def test_repository(self, repo_name: str, valid: bool) -> None:
+        result = re.match(_uri_parts["repository"], repo_name + "/")
+        if valid:
+            assert result is not None
+        else:
+            assert result is None
+
+    @pytest.mark.parametrize(
+        "refexp, valid",
+        [
+            ("", False),
+            ("main", True),
+            ("main@", True),
+            ("main~", True),
+            ("main^", True),
+            ("main^2", True),
+            ("main^^^", True),
+            ("main^1^1", True),
```

Comment:
So `^N^N` is supported? Cool, did not know that, but it does make sense.

Reply:
Note that I have not tried it against a lakeFS server, but the [docs](https://docs.lakefs.io/understand/model.html#ref-expressions) say that it should be supported:

> If `<ref>` is a ref expression, then:
> - [...]
> - `<ref>^N` is a ref expression referring to its N’th parent; in particular `<ref>^1` is the same as `<ref>^`.

Since this a recursive grammar, the repeated `^N` suffix should be accepted as a valid ref expression.

Reply:
Yeah, the caret `^` controls the ancestral depth, and the subsequent number gives the ancestor number. I think this works just like in git.